### ISSUE TITLE
Measure Tool: Does not take SSF in count

### ISSUE THREAD

**Issue by 7xbi** [ESV-Sweetplum/plumoguSV#2]
Would be useful if it did take this in count for SV distance. Idk how hard it is to calculate it in

**Comment by ESV-Sweetplum** [comment]
## Technically possible, but very difficult. Long-winded mathematical explanation below.

Let's start with how measure actually works. Since it will be helpful to draw similarities to real life, I will use a standard formula for relating distance and velocity (in Quaver's case, **scroll velocity**) using an integral:
```math
\Delta x=\int_{t_1}^{t_2} v(t)\,dt
```
Of course, this only applies if the velocity function is continuous. In Quaver's case, we know it isn't, as placing an SV instantly changes the entire chart's speed from one value to another. We could either choose to represent the distance using the integral of various Heaviside functions, where:
```math
\Delta x=\int_{t_i}^{t_f} v\cdot \theta_c(t)
```
In practice, we will use a Riemann sum to compute the distance travelled within a given timeframe:
```math
\Delta x = \sum_{i}^n f(t_i)\Delta t_i
```
where $f(t_i)=$`map.ScrollVelocities[i].Multiplier`, and $\Delta t_i=$`map.ScrollVelocities[i + 1].StartTime - map.ScrollVelocities[i].StartTime`. This is fine for all intents and purposes.

However, when scroll speed factor comes in, things become a bit of an issue, because the entire idea of distance is warped. Not only does the distance travelled by a note across SSFs change depending on the note's temporal (and by relation spatial) location, but the entire idea of velocity changes as well. Think of it like stretching the entirety of space; you're redefining the measure we use for distance. The real life version of this is stretching a meter stick outwards/inwards, and defining that as the new meter.

Since SSF directly affects position (while SV only affects velocity, which is position over TIME), SSFs in game are linearized. Say we did try to find the distance between two SSFs with no SVs. Since the linearization is continuous, we can find the equation of a line between the points $(t_a,S_a)$ and $(t_b,S_b)$, and we will revert to using an integral (note that this integral only represents the movement over time due to the scroll speed factor changes, and not the movement done through changing the SSF itself):
```math
\Delta x = \int_{t_a}^{t_b} S(t)\,dt=\int_{t_a}^{t_b} \frac{\Delta S}{\Delta t}(t-t_a)+S_a\,dt = \frac{1}{2}\frac{dS}{dt}(\Delta t)^2+S_a\Delta t \approx  \left(\frac{1}{2}S_a+S_b\right)\Delta t
```
where $\frac{dS}{dt}=\frac{S_b-S_a}{t_b-t_a}$. Now, this number is quite reasonable, but it tells us two things:
- While the speed between two SVs is constant, the speed between two SSFs is not. This is due to the linearization of SSF, which is necessary to make it not choppy.
- The distance changes quadratically with time instead of linearly, which makes calculating things much more difficult.

It does not really make sense to find a distance **by constantly redefining what a distance is**, which is effectively what scroll speed factor is doing. If I really wanted to (I don't), I could extend these equations into having SV as well, such that the distance between two SVs with changing SSF would be:
```math
\Delta x = \sum_{i}^n v(t_i)\int_{t_a}^{t_b} S(t)\,dt= \sum_{i}^n v(t_i)\left(  \frac{1}{2}\frac{dS}{dt}(\Delta t_i)^2+S_a\Delta t_i \right)
```
where the precise integral version of SSF is required because the SVs may not start at the same time as the SSFs. Yet, how will you use this? Because of the fact that this spits out a number in `msx`, the number you obtain will have to be multiplied by the current given SSF, as an `msx` is now defined as `ms * x * SSFx`, where `SSFx` is a unitless assignment, similar to radians or degrees. If you change any of the displacement SVs, the measure will change in unpredictable ways due to the SSF conflicts. Remember that with all this, the distance that the note travels CANNOT be equal to the above, as remember that the notes' position changes as well due to SSF; this is just representing the movement alone. We could solve this by representing the SSF as the average value of a solution to a differential equation instead:
```math
\frac{dx}{dt} = v(t)\cdot \frac{dS}{dt}, x(0)=x_i
```
But honestly I really don't want to do this and it will take so long for the plugin to actually like... do anything. Rant over.


## TL;DR: It is possible to obtain a number, but the number is somewhat meaningless, and it's not worth the effort implementing something that will take so long computationally. I could do it, but I don't really want to.

Let me know if anything in here is wrong because I confused myself like 5 times writing this

**Comment by ESV-Sweetplum** [comment]
If both SSFs and SVs were discrete, or both of them were continuous, it'd work out a lot better.

**Comment by 7xbi** [comment]
bro..............

**Comment by 7xbi** [comment]
![images](https://github.com/user-attachments/assets/8efabcb2-662b-41b3-827f-d96c9f48184e)

**Comment by BakersDozenBagels** [comment]
I don't think measuring the total distance over a range of SSFs even makes sense as a concept, since changing it twice in opposite ways always has a net 0 effect on distance (unlike SVs, where this can result in e.g. teleports). However, you could calculate the distance including one point's SSF (e.g. start of measurement or end of measurement) with a simple multiplication by the current SSF at that point.

**Comment by ESV-Sweetplum** [comment]
Yeah that's true, probably links back to the fact that SSF is conservative (NOT IN THE POLITICAL WAY)

**Comment by ESV-Sweetplum** [comment]
> I don't think measuring the total distance over a range of SSFs even makes sense as a concept, since changing it twice in opposite ways always has a net 0 effect on distance (unlike SVs, where this can result in e.g. teleports).

You still definitely can represent it mathematically though, because changing it twice in different directions means that the resulting space is identical to the initial space, meaning that d=vt once again

**Comment by ESV-Sweetplum** [comment]
Wait sorry I'm having a realization

If SSF changes are conservative, then $\int_a^b f(t)\,dt = F(b)-F(a)$ applies.... let me take more time to think about this

**Comment by BakersDozenBagels** [comment]
With note position as $P$ and at a given constant time:
$SSF \propto P$
$SV \propto \frac{dP}{dt} \rightarrow \int SV dt \propto P$
You simply don't integrate SSF.

**Comment by ESV-Sweetplum** [comment]
Hmm yeah your point makes sense, the whole conservative thing was a nice realization to the fact that you can just skip the integral entirely and evaluate at point A and point B

**Comment by ESV-Sweetplum** [comment]
The current measure tool does not account for SSF in any capacity, so here's the plan:

Let $x_i=\text{measure(startOffset, desiredNoteOffset)}$

Then the outputted distance between two SVs will be $d=x_{i}\left(S_{a}-S_{b}\right)+vt_{total}S_{b}$. 

For the entire region:
where there are ssfs, place hypothetical svs of the previous sv value.
between every sv, linearly interpolate from most recent and next ssf. these two values will be used for s_a and s_b.
calculate distance, repeat for every sv, sum distances
remove hypothetical svs

i will start working on this now 

Slight issue; $x_i$ needs to be recomputed per sv, meaning we're entering O(n^2)...

**Comment by ESV-Sweetplum** [comment]
tried for a while to make an implementation... didn't really work out well because propagating errors would make the end result kinda useless

closing this for now, if someone else is willing to implement it open a PR
